### ISSUE TITLE
Field(..., exclude=True) causes SchemaError if model is later referenced again

### ISSUE THREAD

**Issue by oefe** [pydantic/pydantic#5457]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2 installed directly from the `main` branch, or equivalent

### Description

When a model references another model via an excluded field (`Field(..., exclude=True)`), and later references that model again (not excluded), Pedantic raises an unexpected `SchemaError`.

This also happens if either reference is indirect (e.g. using a `list[Model]`).

The error doesn't happen if the excluded field comes after the not excluded field(s). 

```
>python slots_error.py
Traceback (most recent call last):
  File "/Users/moefelei/Repos/pydantic-sandbox/slots_error.py", line 8, in <module>
    class UpdateTagListRequest(BaseModel):
  File "/Users/moefelei/Repos/pydantic-sandbox/.venv/lib/python3.11/site-packages/pydantic/main.py", line 149, in __new__
    _model_construction.complete_model_class(
  File "/Users/moefelei/Repos/pydantic-sandbox/.venv/lib/python3.11/site-packages/pydantic/_internal/_model_construction.py", line 190, in complete_model_class
    cls.__pydantic_serializer__ = SchemaSerializer(schema, core_config)
                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
pydantic_core._pydantic_core.SchemaError: Error building `model` serializer:
  SchemaError: Error building `typed-dict` serializer:
  SchemaError: Field `tag`:
  SchemaError: Error building `definition-ref` serializer:
  SchemaError: Slots Error: ref '__main__.Tag:140369789686544' not found
```

### Example Code

```Python
from pydantic import BaseModel, Field


class Tag(BaseModel):
    name: str


class UpdateTagListRequest(BaseModel):
    excluded_list: list[Tag] = Field(..., exclude=True)
    tag: Tag
```

### Python, Pydantic & OS Version

```Text
pydantic version: 2.0a1
                 install path: /Users/moefelei/Repos/pydantic-sandbox/.venv/lib/python3.11/site-packages/pydantic
               python version: 3.11.2 (main, Feb 16 2023, 03:07:35) [Clang 14.0.0 (clang-1400.0.29.202)]
                     platform: macOS-13.2.1-x86_64-i386-64bit
     optional deps. installed: ['typing-extensions']
```

**Comment by oefe** [comment]
I pinned pydantic-core at 0.22.0, but it also happens with pydantic-core 0.23.1

**Comment by dmontagu** [comment]
Thanks for reporting, I can confirm it happens on latest commits of pydantic/pydantic_core

**Comment by samuelcolvin** [comment]
nice, thanks for reporting.

**Comment by dmontagu** [comment]
This seems to be resolved in v2.0b3, so I'm going to close this issue, but please let us know if you run into any other related issues.
